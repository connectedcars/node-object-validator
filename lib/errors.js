function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
