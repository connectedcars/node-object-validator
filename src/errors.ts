export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
