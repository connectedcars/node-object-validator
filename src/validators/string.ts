import { CodeGenResult, ValidatorBase, ValidatorOptions } from '../common'
import { NotStringFail, RequiredFail, ValidationErrorContext, ValidationFailure, WrongLengthFail } from '../errors'

export function isString(
  value: unknown,
  minLength = 0,
  maxLength: number = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): value is string {
  const errors = validateString(value, minLength, maxLength, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateString(
  value: unknown,
  minLength = 0,
  maxLength: number = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): ValidationFailure[] {
  if (typeof value !== 'string') {
    return [new NotStringFail(`Must be a string (received "${value}")`, context)]
  }
  if ((minLength !== 0 && value.length < minLength) || value.length > maxLength) {
    return [
      new WrongLengthFail(
        `Must contain between ${minLength} and ${maxLength} characters (received "${value}")`,
        context
      )
    ]
  }
  return []
}

export class StringValidator<O = never> extends ValidatorBase<string | O> {
  private minLength: number
  private maxLength: number

  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(options)
    this.minLength = minLength
    this.maxLength = maxLength
    if (options?.optimize) {
      this.optimize()
    }
  }

  public validate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
    if (value == null) {
      return this.required ? [new RequiredFail(`Is required`, context)] : []
    }
    return validateString(value, this.minLength, this.maxLength, context)
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: ValidationErrorContext
  ): CodeGenResult {
    const contextStr = context ? `, { key: \`${context.key}\` }` : ', context'
    const localValueRef = `value${id()}`
    const declarations: string[] = []
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
      `if (${localValueRef} != null) {`,
      `  if (typeof ${localValueRef} === 'string') {`,
      `    if (${this.minLength ? `${localValueRef}.length < ${this.minLength} || ` : '' }${localValueRef}.length > ${this.maxLength}) {`,
      `      errors.push(new WrongLengthFail(\`Must contain between ${this.minLength} and ${this.maxLength} characters (received "\${${localValueRef}}")\`${contextStr}))`,
      `    }`,
      `  } else {`,
      `    errors.push(new NotStringFail(\`Must be a string (received "\${${localValueRef}}")\`${contextStr}))`,
      `  }`,
      ...(this.required ? [
        `} else {`,
        `  errors.push(new RequiredError(\`Is required\`${contextStr}))`] : []),
        '}'
    ]
    return [
      {
        WrongLengthFail: WrongLengthFail,
        NotStringFail: NotStringFail,
        RequiredError: RequiredFail
      },
      declarations,
      code
    ]
  }
}

export class RequiredString extends StringValidator {
  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(minLength, maxLength, { ...options, required: true })
  }
}

export class OptionalString extends StringValidator<undefined | null> {
  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(minLength, maxLength, { ...options, required: false })
  }
}
