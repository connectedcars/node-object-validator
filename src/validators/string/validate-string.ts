import { NotStringError, ValidationErrorContext, WrongLengthError } from '../../errors'

export function validateString(
  value: string,
  minLength = 0,
  maxLength: number = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext<string>
): Error | null {
  if (typeof value !== 'string') {
    return new NotStringError(`Must be a string (received "${value}")`, context)
  }
  if ((minLength !== 0 && value.length < minLength) || value.length > maxLength) {
    return new WrongLengthError(
      `Must contain between ${minLength} and ${maxLength} characters (received "${value}")`,
      context
    )
  }
  return null
}
