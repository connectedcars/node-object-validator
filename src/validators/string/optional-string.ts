import { ValidationErrorContext } from '../../errors'
import { validateString } from './validate-string'

export class OptionalString {
  private type: 'OptionalString' = 'OptionalString'
  private minLength: number
  private maxLength: number

  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: string, context?: ValidationErrorContext<string>): Error | null {
    if (value == null) {
      return null
    }
    return validateString(value, this.minLength, this.maxLength, context)
  }
}
