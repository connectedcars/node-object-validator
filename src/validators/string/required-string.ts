import { RequiredError, ValidationErrorContext } from '../../errors'
import { validateString } from './validate-string'

export class RequiredString {
  private type: 'RequiredString' = 'RequiredString'
  private minLength: number
  private maxLength: number

  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: string, context?: ValidationErrorContext<string>): Error | null {
    if (value == null) {
      return new RequiredError(`Is required`, context)
    }
    return validateString(value, this.minLength, this.maxLength, context)
  }
}
