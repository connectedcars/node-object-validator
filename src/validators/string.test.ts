import { NotStringError, RequiredError, WrongLengthError } from '../errors'
import { OptionalString, RequiredString, validateString } from './string'

describe('String', () => {
  describe('RequiredString', () => {
    it('requires empty value', function() {
      const validator = new RequiredString()
      expect(validator.validate(null)).toStrictEqual([new RequiredError('Is required')])
      expect(validator.validate(undefined)).toStrictEqual([new RequiredError('Is required')])
    })
  })

  describe('OptionalString', () => {
    it('requires empty value', function() {
      const validator = new OptionalString()
      expect(validator.validate(undefined)).toStrictEqual(null)
      expect(validator.validate(undefined)).toStrictEqual(null)
    })
  })

  describe('validateString', () => {
    it('requires value to be a string', () => {
      expect(validateString('foo')).toStrictEqual(null)
      expect(validateString('')).toStrictEqual(null)
      expect(validateString((1 as unknown) as string)).toStrictEqual([
        new NotStringError('Must be a string (received "1")')
      ])
      expect(validateString(({} as unknown) as string)).toStrictEqual([
        new NotStringError('Must be a string (received "[object Object]")')
      ])
      expect(validateString(([] as unknown) as string)).toStrictEqual([
        new NotStringError('Must be a string (received "")')
      ])
      expect(validateString((true as unknown) as string)).toStrictEqual([
        new NotStringError('Must be a string (received "true")')
      ])
      expect(validateString((false as unknown) as string)).toStrictEqual([
        new NotStringError('Must be a string (received "false")')
      ])
    })

    it('requires min value length', () => {
      expect(validateString('', 5, 500)).toStrictEqual([
        new WrongLengthError('Must contain between 5 and 500 characters (received "")')
      ])
      expect(validateString('a', 5, 500)).toStrictEqual([
        new WrongLengthError('Must contain between 5 and 500 characters (received "a")')
      ])
      expect(validateString('ab', 5, 500)).toStrictEqual([
        new WrongLengthError('Must contain between 5 and 500 characters (received "ab")')
      ])
      expect(validateString('abc', 5, 500)).toStrictEqual([
        new WrongLengthError('Must contain between 5 and 500 characters (received "abc")')
      ])
      expect(validateString('abcd', 5, 500)).toStrictEqual([
        new WrongLengthError('Must contain between 5 and 500 characters (received "abcd")')
      ])
      expect(validateString('abcde', 5, 500)).toStrictEqual(null)
      expect(validateString('abcdef', 5, 500)).toStrictEqual(null)
      expect(validateString('this is a long string', 5, 500)).toStrictEqual(null)
    })

    it('requires max value length', () => {
      expect(validateString('', 0, 5)).toStrictEqual(null)
      expect(validateString('a', 0, 5)).toStrictEqual(null)
      expect(validateString('ab', 0, 5)).toStrictEqual(null)
      expect(validateString('abc', 0, 5)).toStrictEqual(null)
      expect(validateString('abcd', 0, 5)).toStrictEqual(null)
      expect(validateString('abcde', 0, 5)).toStrictEqual(null)
      expect(validateString('abcdef', 0, 5)).toStrictEqual([
        new WrongLengthError('Must contain between 0 and 5 characters (received "abcdef")')
      ])
      expect(validateString('abcdefg', 0, 5)).toStrictEqual([
        new WrongLengthError('Must contain between 0 and 5 characters (received "abcdefg")')
      ])
      expect(validateString('this is a long string', 0, 5)).toStrictEqual([
        new WrongLengthError('Must contain between 0 and 5 characters (received "this is a long string")')
      ])
    })
  })
})
