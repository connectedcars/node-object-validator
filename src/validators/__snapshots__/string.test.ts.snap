// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String (optimize: %s) RequiredString should return an function body 1`] = `
Array [
  Object {
    "NotStringFail": [Function],
    "RequiredFail": [Function],
    "WrongLengthFail": [Function],
  },
  Array [],
  Array [
    "const value1 = value1",
    "if (value1 !== undefined) {",
    "  if (typeof value1 === 'string') {",
    "    if (value1.length > 10) {",
    "      errors.push(new WrongLengthFail(\`Must contain between 0 and 10 characters\`, value1, context))",
    "    }",
    "  } else {",
    "    errors.push(new NotStringFail(\`Must be a string\`, value1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
